Mark Framework Interfaces as API part 3

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

The following interfaces marked as API

lib/internal/Magento/Framework/Async/CancelableDeferredInterface.php
lib/internal/Magento/Framework/Async/DeferredInterface.php
lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php
lib/internal/Magento/Framework/Backup/SourceFileInterface.php
lib/internal/Magento/Framework/Cache/ConfigInterface.php
lib/internal/Magento/Framework/Cache/StaleCacheNotifierInterface.php
lib/internal/Magento/Framework/Code/Generator/CodeGeneratorInterface.php
lib/internal/Magento/Framework/Code/Minifier/AdapterInterface.php
lib/internal/Magento/Framework/Code/Reader/ClassReaderInterface.php
lib/internal/Magento/Framework/Code/ValidatorInterface.php
lib/internal/Magento/Framework/Component/ComponentRegistrarInterface.php
lib/internal/Magento/Framework/Crontab/CrontabManagerInterface.php
lib/internal/Magento/Framework/Crontab/TasksProviderInterface.php
lib/internal/Magento/Framework/Css/PreProcessor/ErrorHandlerInterface.php
lib/internal/Magento/Framework/DB/DataConverter/DataConverterInterface.php
lib/internal/Magento/Framework/DB/LoggerInterface.php
lib/internal/Magento/Framework/DB/MapperInterface.php
lib/internal/Magento/Framework/DB/Query/BatchIteratorInterface.php
lib/internal/Magento/Framework/DB/QueryInterface.php
lib/internal/Magento/Framework/DB/Select/QueryModifierInterface.php
lib/internal/Magento/Framework/DB/Select/RendererInterface.php
lib/internal/Magento/Framework/DB/Sequence/SequenceInterface.php
lib/internal/Magento/Framework/DB/Sql/ExpressionInterface.php
lib/internal/Magento/Framework/Data/Collection/Db/FetchStrategyInterface.php
lib/internal/Magento/Framework/Data/Collection/EntityFactoryInterface.php
lib/internal/Magento/Framework/Data/CollectionDataSourceInterface.php
lib/internal/Magento/Framework/Data/CollectionModifierInterface.php
lib/internal/Magento/Framework/Data/SearchResultInterface.php
lib/internal/Magento/Framework/Data/SearchResultProcessorInterface.php
lib/internal/Magento/Framework/Data/ValueSourceInterface.php
lib/internal/Magento/Framework/DataObject/IdentityGeneratorInterface.php
lib/internal/Magento/Framework/DataObject/IdentityInterface.php
lib/internal/Magento/Framework/DataObject/IdentityValidatorInterface.php
lib/internal/Magento/Framework/DataObject/KeyValueObjectInterface.php

Marked api as references
lib/internal/Magento/Framework/Api/CriteriaInterface (by lib/internal/Magento/Framework/DB/MapperInterface.php)

Should merge this before PR related below to avoid conflicts same file

### Dependent Pull Requests
[magento/magento2ce#32135](https://github.com/magento/magento2/pull/32135)

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#32026

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

mrtuvn commented:
rebased branch 2.5-develop

mrtuvn commented:
@magento run all tests

mrtuvn commented:
Public type 'Magento\Framework\DB\MapperInterface' references following non-public types:
Magento\Framework\Api\CriteriaInterface

mrtuvn commented:
Fixed static tests fail

mrtuvn commented:
@magento run all tests

eduard13 reviewed (changes requested):
@mrtuvn please revert the `api` annotation for `Magento/Framework/Api/CriteriaInterface` as it has been already marked in scope of #32135. Let's add it as related PR.

Additionally, could you please enable the strict typing for the adjusted classes?

eduard13 commented:
@mrtuvn please update the PR description (for some reasons I don’t have the permission), Related Pull Requests section with the PR that I mentioned, so it will be used on building the environment.

mrtuvn commented:
Added related PR

eduard13 commented:
@magento run Static Tests

eduard13 commented:
@mrtuvn please add it according to the following [doc](https://devdocs.magento.com/contributor-guide/maintainer-handbook.html#related-pull-requests).

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

eduard13 commented on lib/internal/Magento/Framework/Autoload/AutoloaderInterface.php:
```diff
@@ -18,7 +22,7 @@ interface AutoloaderInterface
      * @return void
      */
-    public function addPsr4($nsPrefix, $paths, $prepend = false);
+    public function addPsr4($nsPrefix, $paths, $prepend = false): void;
```
@mrtuvn it's not a good idea to specify the returning type, as it will fail all the implementations. Let's only add `declare(strict_types=1)`, as you've already did, and should be enough for now.

mrtuvn replied:
so I don't need add return type after every method right ? Only add strict at first line 
Just a bit confuse about consistency across entire ecosystem. While somewhere I see that type return and somewhere it's omitted

eduard13 replied:
> so I don't need add return type after every method right ? Only add strict at first line

Exactly.

Also, looks like linking PR (Related PR) didn't work. Could you please double check that you use exactly the same way? The placeholder should also be there.

![image](https://user-images.githubusercontent.com/15868188/108813929-a1920400-75ba-11eb-8360-5432bf03c0c4.png)

mrtuvn commented:
Corrected it @eduard13

eduard13 commented:
@magento run Static Tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
added related PR but seem not solve problem with fail static tests

eduard13 commented:
@mrtuvn most likely the PR wasn't added correctly as related. The related PR should be displayed here if everything is correctly added:
![image](https://user-images.githubusercontent.com/15868188/108965428-86d89180-7685-11eb-8b96-5f39d7bac190.png)
One more `CE` row should be displayed, pointing to that PR.

Please try multiple variations of adjusting that section, and check the Build Configuration.

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
CC: @slavvka @sidolov @sivaschenko Any ideas what happen with build configuration ? I got fails static when already set related pull requests (maybe wrong syntax) ?

eduard13 commented:
Hi @mrtuvn, after an internal discussion, let's simply adjust the classes, by marking them as `api`, as you initially did. Sorry for the confusion, but looks like relating PRs from the same repo isn't yet supported.

Thank you.

eduard13 commented on lib/internal/Magento/Framework/DB/QueryInterface.php:
```diff
@@ -44,7 +48,7 @@ public function getSize();
      *
      * @param bool $stringMode
-     * @return string || Select
+     * @return string|Select
```
I’d also revert such adjustments, because of which the SVC is failing, and requires a separate approval step.

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run WebAPI Tests, Unit Tests, Semantic Version Checker, Sample Data Tests EE, Sample Data Tests CE, Sample Data Tests B2B, Magento Health Index, Integration Tests, Functional Tests EE, Functional Tests CE, Functional Tests B2B, Database Compare

mrtuvn commented:
@magento run WebAPI Tests,  Semantic Version Checker, Integration Tests

eduard13 commented:
@magento run Integration Tests

mrtuvn commented:
@magento run Semantic Version Checker

eduard13 commented:
@mrtuvn as I mentioned above, the SVC is failing because of touching the return annotations of some API methods.

mrtuvn committed:
Mark Framework Interfaces as API part 3


fix static


update


update


fix non-public references 


revert change return value interface

mrtuvn commented:
Yes i have corrected that! Thanks you for the information
@magento run all tests

eduard13 commented:
@magento run Sample Data Tests EE

magento-engcom-team commented:
Hi @eduard13, thank you for the review.
[ENGCOM-8833](https://jira.corp.magento.com/browse/ENGCOM-8833) has been created to process this Pull Request

mrtuvn commented:
Re-targeted PR branch to 2.4-develop as internal discussed with @gabrieldagama

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests CE, Functional Tests B2B

gabrieldagama commented:
@magento import code to https://github.com/magento-engcom/magento2ce

magento-engcom-team commented:
@gabrieldagama the branch with code successfully imported into`magento-engcom/magento2ce` repository. Branch name: `imported-magento-magento2-32156`.
